### PR TITLE
Create target base class

### LINKED ISSUE #67: Target Class
Base class for an entire hierarchy of classes. Basic API is found [here](https://raw.githubusercontent.com/CMakePP/CMaize/master/docs/src/developer/overall_design.png).

Modern CMake is target based meaning for CMaize to be able to interact with CMake effectively CMaize will need to be able to generate CMake targets. The `Target` class provides an object-oriented API for interacting with a CMake target. It also serves as code factorization for functionality/APIs common to all
target types. Native CMake only has one type of target, which can be annoying since the target is interacted with in different ways depending on what it describes. The classes which derive from `Target` provide a more user-friendly means of interacting with specific target types, than native CMake does.

### PR DESCRIPTION
**Pull request by zachcran**
Depends on PRs #76 and #77. Closes #67. This PR will create a base class for the class hierarchy representing various types of CMake targets.

Todos:
- [x] Write the class
- [x] Test the class
- [x] Write design notes
- [x] Verify docs are complete and build

### CODE REVIEW

File: cmake/cmaize/cmaize.cmake [hunk 1]
```diff
@@ -11,4 +11,8 @@ if("${CMAKE_BINARY_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
     message(FATAL_ERROR "${_msg}")
 endif()
 
+include(cmaize/project_specification/project_specification)
+include(cmaize/targets/targets)
 include(cmaize/toolchain/toolchain)
+include(cmaize/project_specification/project_specification)
```
Comment (ryanmrichard) [inline cmake/cmaize/cmaize.cmake:17]: The include shows up twice.
```suggestion
include(cmaize/project_specification/project_specification)
include(cmaize/targets/targets)
include(cmaize/toolchain/toolchain)
```

Comment (zachcran) [reply cmake/cmaize/cmaize.cmake:17]: Right, I forgot to fix that after fixing the merge conflict. Thank you!

**Comment by zachcran** [comment]
That conflict keeps popping back up... hopefully it is fixed now for good.

**Comment by zachcran** [comment]
@ryanmrichard r2g

File: cmake/cmaize/targets/target.cmake [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    # :param self: Target object
+    # :type self: Target
+    # :param has_property: Return variable for if the target has the property.
+    # :type has_property: bool
+    # :param property_name: Name of the property to check for.
+    # :type property_name: str
+    #
+    # :returns: Sets ``has_property`` according to if the target has the
+    #           requested property (True) or not (False).
+    # :rtype: bool
```
Comment (ryanmrichard) [inline cmake/cmaize/targets/target.cmake:100]:
```suggestion
    # :rtype: bool*
```
FYI https://cmakepp.github.io/CMakePPLang/features/types.html#pointers is how I recommend documenting the return type.

**Review by ryanmrichard — DISMISSED** [review]

File: cmake/cmaize/targets/target.cmake (hunk 2 above)
Comment (zachcran) [reply cmake/cmaize/targets/target.cmake:100]: Would you recommend using the pointer typing for all return variables?

File: cmake/cmaize/targets/target.cmake [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    #]]
+    cpp_member(has_property Target desc str)
+    function("${has_property}" self has_property property_name)
+
+        Target(target "${self}" my_name)
+
+        # Call CMake's built-in `get_target_property()` method.
+        get_target_property(property_value "${my_name}" "${property_name}")
```
[the remark this answers is not in the document]
Comment (zachcran) [reply cmake/cmaize/targets/target.cmake:108]: Huh, I didn't know that CMake would crash for certain target/property combinations. I am imagining this as an internally managed whitelist in the Target class, with children populating it with needed values. Then, the property method calls from Target will just need to check that whitelist as necessary before doing their operations. Does that sound about right?

File: cmake/cmaize/targets/target.cmake (hunk 2 above)
Comment (ryanmrichard) [reply cmake/cmaize/targets/target.cmake:100]: yep

File: cmake/cmaize/targets/target.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanmrichard) [reply cmake/cmaize/targets/target.cmake:97]: IIRC I ran into this a lot with imported targets. If it were me I'd probably just make a free function along the lines of:
```
is_allowed_property(result target_type property_name)
```
you could then call `is_allowed_property(result IMPORTED INTERFACE_LINK_LIBRARIES)` and find out if the call `get_target_property(property_value <imported_target_name> INTERFACE_LINK_LIBRARIES)` is allowed. 

Unfortunately unless someone's published the list you're going to have to reverse engineer which properties are allowed and which aren't, as of last check, CMake doesn't publish such a list.

**Review by ryanmrichard — APPROVED** [review]

**Pull request merged**